Get speed

Fildo7525 opened the pull request:
Odometria robota. Pocitanie pozicie robota  na zaklade rychlosti, ktore boli vyziadane od robota

Commit:
odometry: failes: get speeds

The robot does not send back the desired json string.
The response is that he received it.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
DO-NOT-MERGE: clear unimportant messages

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: The first response is if the request was received

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: changed logging levels

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: be sure to create the files

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: show if the client connected

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: QUESTION: get the speeds of the motors

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: calculate the robot location

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: calculation the robot position

The calculation calculates the x, y coordinates of the robot
and its angle of rotation.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: change debugging

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: measure the send and receive time

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cp: reorganizing the includes

The includes are in this order:
    1. The header file
    2. Files from this project
    3. System files
    4. ROS2 files
    5. Cpp STL files

Includes are ordered in alphabetical order.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
deprication: do not use getStoppedTimes

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
logs: Created merged log file from files of your choosing

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
README: updated

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
doxygen: include private methods to the documentation

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: add comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: limit the received velocities

When the robot does not move. The received messages contain excessively
massive numbers. The maximal angular speed that can be created on the
motor's shaft is limited by the software to 1000 rps.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: the position is calculated using doubles

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: TODO

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: stopwatch: use stopwatch on every build type

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: measure the whole preparation code

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: synchronization on send and receive

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: typo fix

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Fildo7525 commented:
rebase na mastra

Commit:
global: reorganize private variables

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
global: fix some log statements

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: display changed velocities on change

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: create logs in separate directories

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: send int's

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: initialize the last stamp of velocity as a maximal value

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: display the message instead of the nnumber of sent bytes

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
logs: backup them before delete

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
FATAL: odometry: change the robot speed

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

MichalDobis commented on src/blackmetal/src/communication/controlSW/Odometry.hpp:
```diff
@@ -0,0 +1,116 @@
+	 * @param jsonMessage Message received from the server.
+	 * @return Structure of left and right wheel speed.
+	 */
+	Speed obtainWheelSpeeds(const std::string &jsonMessage);
```
povedal by som, zeby asi mohla byt const
```suggestion
	Speed obtainWheelSpeeds(const std::string &jsonMessage) const;
```

MichalDobis commented on src/blackmetal/src/communication/controlSW/Odometry.hpp:
```diff
@@ -0,0 +1,116 @@
+	 * @param speed Rvalue reference to a structure containing the left and right wheel velocity.
+	 * @param elapsedTime The rvalue reference to the elapsed time after the receiving the robot velocity. 
+	 */
+	void changeRobotLocation(Speed &&speed, long double &&elapsedTime);
```
co je robot location?

Fildo7525 replied:
je to pozicia robota v kartezianskej sustave, ked jeho uvodna pozicia po zapnuti programu je [0, 0, 0] (x, y, natocenie po z-osi)

MichalDobis commented on src/blackmetal/src/communication/controlSW/Odometry.hpp:
```diff
@@ -0,0 +1,116 @@
+
+private:
+	/// Instance of the control software client.
+	BlackMetal &m_controlSoftware;
```
nebol by lepsi ```std::unique_ptr```?

Fildo7525 replied:
unique_pointer pokial viem tak sa pouziva ked chces mat len jedneho ownera pre alokovany objekt. V tomto pripade ja nealokujem ziadne miesto pre tuto klasu lebo je to alokovane rosom a to uz je dane ako shared_pointer takze by to ist nemalo. Ale neskusal som to.

MichalDobis replied:
```BlackMetal &m_controlSoftware``` toto je ale referencia nie? Myslel som, ze ownerom je iba tato class. Ak je aj ina tak by tam mal byt ```std::shared_ptr```

Fildo7525 replied:
spravene

MichalDobis commented on src/blackmetal/src/communication/controlSW/Odometry.cpp:
```diff
@@ -0,0 +1,171 @@
+	double changeOfAngleInTime = angularVelocity * dt;
+	DBG("angle change: " << changeOfAngleInTime);
+
+	{
+		std::lock_guard<std::mutex> lock(g_robotLocationMutex);
+		m_coordination.x += dxt;
+		m_coordination.y += dyt;
+		m_coordination.angle += changeOfAngleInTime;
+	}
+}
```
tu uz nepotrebujes robit scope, lock ti skonci na konci funkcii
```suggestion

	std::lock_guard<std::mutex> lock(g_robotLocationMutex);
	m_coordination.x += dxt;
	m_coordination.y += dyt;
	m_coordination.angle += changeOfAngleInTime;
}
```

Fildo7525 replied:
Hej robit to nepotrebujem ale je to na znak ze to tam je a je to viac  viditelne v kode. Nakoniec na tom az tak nezalezi, dal som to tak kvoli citatelnosti a kvoli tomu ze to tak robim vsade v kode a tymto sposobom len dodrzujem konzistenciu pisania kodu.

MichalDobis replied:
:+1:

MichalDobis commented on src/blackmetal/src/communication/controlSW/Odometry.cpp:
```diff
@@ -0,0 +1,171 @@
+	return bm::Status::OK;
+}
+
+void Odometry::changeRobotLocation(Speed &&speed, long double &&elapsedTime)
```
nerozumiem stale tejto funkcii. Odometria by mala mat vystup rychlost a polohu a ako vstup TCP/IP clienta alebo raw data. A nerozumiem preco by mala menit nejaku lokalitu

Fildo7525 replied:
meni lokaciu robota v kartezianskej sustave. Kedze je client spravovany triedou BlackMetal tak si vytvorim novy thread vo vytvorenom objekte a v nom pollujem robota a ziskavam jeho rychlosti. Tie nasledne presuvam, nekopirujem, v strukture 'Speed' spolu s casom, ktory trvala komunikacia so serverom, Na zaklade dodanych informacii prepocitavam polohu tohto robota v systeme.

MichalDobis replied:
aha uz rozumiem, mozno by bolo lepsie keby bol client zdielany medzi BlackMetal a touto triedou

Fildo7525 committed:
odometry: get the wheel speed from const class

Co-authored-by: MichalDobis <miso.dobis@gmail.com>

Commit:
stopwatch: The functionality is needed int the release verion, too

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: thread safe queue

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: get an argument as a shared pointer

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

MichalDobis commented on src/blackmetal/src/communication/queue/Queue.hpp:
```diff
@@ -0,0 +1,35 @@
+namespace ts
+{
+
+class Queue {
```
👍

Commit:
client: execute updated

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

MichalDobis commented on src/blackmetal/src/communication/queue/Queue.cpp:
```diff
@@ -0,0 +1,54 @@
+	bool full = m_maxSize == m_queue.size();
+
+	if (m_queue.empty()) {
+		return std::move(tmp);
```
mozno by bol lepsi ```throw EmptyQueueException()```

Fildo7525 replied:
mozem , ale ako tymto som zabezpecit ze to zoberie hned ako tam nieco bude.

MichalDobis replied:
no to by mala tam condition variabla oznamit, ze sa tam nieco pridalo.

Commit:
queue: wait for pushing and popping

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: share a client with balckmetal

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: priority queue

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: worker thread

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: add comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: logging with mutex

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: max speed limited by rpm

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: remove unused code

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: use the evaluation internally

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cp: odometry: typos

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: Use priority queue for sending and receiving

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: return data from future variable

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
bm: set the restrictions for the speed

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: apply the send principal to receive

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: run the workerThread in detached thread

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: set timer for send and receive

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: send and receive are protected

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cp: rename execute to sendRequest

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: log with thread id

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: give names to queues for debugging

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
bmlogger: cmake: link queue

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: fix queue execution

When you start the program everything works fine.
However when you send a command to server the queue gets stuck.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
FATAL: the condition variable blocks the system

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: log also all to one file

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: the ability to turn off the loggin for a module

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: maybe possible fix

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
mutexes: Apply new standard

According to the new standard the std::scoped_lock should be preferred.
It does exactly the same thing as std::lock_guard, although it takes
variadic number of mutexes and locks them at the same time. Thus
preventing the deadlock.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: mutexes are not needed any more due to the queue

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cp: commets: actualize the documentation

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
doc: Readme and doxygen config update

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
better be safe

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
test2

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: set getSpeedCommand flag to false after sending it

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
test

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: size and empty cannot be const because of mutex

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: odometry as a ts::Queue

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: initialize worker thread with copy of current data

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
logger: reorganize methods

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>
